src/scoped.rs

```diff
@@ -146,47 +144,58 @@ pub unsafe fn builder_spawn_unsafe<'a, F>(
 where
     F: FnOnce() + Send + 'a,
 {
-    use std::mem;
-
-    let closure: Box<FnBox + 'a> = Box::new(f);
-    let closure: Box<FnBox + Send> = mem::transmute(closure);
+    let closure: Box<FnBox<()> + 'a> = Box::new(f);
+    let closure: Box<FnBox<()> + Send> = mem::transmute(closure);
     builder.spawn(move || closure.call_box())
 }
 
-
 pub struct Scope<'a> {
-    dtors: RefCell<Option<DtorChain<'a>>>,
+    dtors: RefCell<Option<DtorChain<'a, ()>>>,
+    joins: RefCell<Option<DtorChain<'a, Option<thread::Result<()>>>>>,
 }
 
-struct DtorChain<'a> {
-    dtor: Box<FnBox + 'a>,
-    next: Option<Box<DtorChain<'a>>>,
+struct DtorChain<'a, T> {
+    dtor: Box<FnBox<T> + 'a>,
+    next: Option<Box<DtorChain<'a, T>>>,
 }
 
-enum JoinState {
-    Running(thread::JoinHandle<()>),
-    Joined,
+impl<'a, T> DtorChain<'a, T> {
+    pub fn pop(chain: &mut Option<DtorChain<'a, T>>) -> Option<Box<FnBox<T> + 'a>> {
+        chain.take().map(|mut node| {
+            *chain = node.next.take().map(|b| *b);
+            node.dtor
+        })
+    }
 }
 
-impl JoinState {
-    fn join(&mut self) {
-        let mut state = JoinState::Joined;
-        mem::swap(self, &mut state);
-        if let JoinState::Running(handle) = state {
-            let res = handle.join();
+struct JoinState<T> {
+    join_handle: thread::JoinHandle<()>,
+    result: usize,
+    _marker: PhantomData<T>,
+}
 
-            if !thread::panicking() {
-                res.unwrap();
-            }
+impl<T> JoinState<T> {
+    fn new(join_handle: thread::JoinHandle<()>, result: usize) -> JoinState<T> {
+        JoinState {
+            join_handle: join_handle,
+            result: result,
+            _marker: PhantomData,
         }
     }
+
+    fn join(self) -> thread::Result<T> {
+        let result = self.result;
```

Comment:
I didn't go through the logic in great detail so maybe this is fine, but on line 430 this value is written into from another thread. Is there definitely no race condition between that and the read here?

Reply:
I added the following doc: "The `Box` constructed below is written only by the spawned thread, and read by the current thread only after the spawned thread is joined (`JoinState::join()`). Thus there are no data races."

Reply:
Thanks!